lib/Conversion/StandardToHandshake/StandardToHandshake.cpp

```diff
@@ -1269,6 +1269,95 @@ void replaceCallOps(handshake::FuncOp f, ConversionPatternRewriter &rewriter) {
   }
 }
 
+/// Rewrite affine.for in a handshake.func into its standard form. Affine
+/// analysis should be carried out before this function being called.
+LogicalResult rewriteAffineFor(handshake::FuncOp f,
+                               ConversionPatternRewriter &rewriter) {
+  // Get all affine.for operations in the function body.
+  SmallVector<Operation *, 8> forOps;
+  f.walk([&](Operation *op) {
+    if (isa<mlir::AffineForOp>(op))
+      forOps.push_back(op);
+  });
+
+  // TODO: how to deal with nested loops?
+  for (unsigned i = 0, e = forOps.size(); i < e; i++) {
+    auto forOp = dyn_cast<mlir::AffineForOp>(forOps[i]);
+
+    // Insert lower and upper bounds right at the position of the original
+    // affine.for operation.
+    rewriter.setInsertionPoint(forOp);
+    auto loc = forOp.getLoc();
+    auto lowerBound = expandAffineMap(rewriter, loc, forOp.getLowerBoundMap(),
+                                      forOp.getLowerBoundOperands());
+    auto upperBound = expandAffineMap(rewriter, loc, forOp.getUpperBoundMap(),
+                                      forOp.getUpperBoundOperands());
+    if (!lowerBound || !upperBound)
```

Comment:
How does this code differ from AffineForLowering?  https://github.com/llvm/llvm-project/blob/master/mlir/lib/Conversion/AffineToStandard/AffineToStandard.cpp#L342

It goes directly to the standard dialect, but is that necessary?  I guess this goes back to the question of how you are planning to get the array dependence information through to the handshake part?

Reply:
The major difference is that, as you mentioned, it directly goes to the standard dialect without the need for SCF. The code size is therefore small.

It is possible to directly go to handshake though, that would take more time to implement, and the current approach doesn't seem to introduce much code here. Please let me know if you think I should go for generating directly to Handshake.

A bit more on how the dependence information is passed: in `rewriteAffineFor` I don't rewrite `affine.load` and `affine.store`, they are handled in `replaceMemoryOps`. And in the affine analysis pass to be implemented, I will get dependence analysis before lowering and store the dependence graph among affine operations, and when lowering them, I also store the mapping between affine load/store and handshake load/store. In this way, I can figure out the dependences among the handshake load/store operations. 

I've sort of implemented this approach in #71 , and theoretically, it should work in this refactorised version.

Reply:
OK, I understand.  I think you need to take some aspects of this description and put it into the code above.  e.g. something like: "This code is similar in many ways to the lowering of Affine loops in AffineToStandard (see [some link](some url)).  The key difference is that this code does not immediately lower affine memory ops as well.   In this code, these will be lowered by rewriteAffineMemoryLocation() taking into account affine memory dependence information."

Reply:
Thank you Steve @stephenneuendorffer , I've updated the description accordingly. Please let me know if it looks fine and whether there is any other things need fix :)